src/rgw/rgw_redis_driver.cc

```diff
@@ -0,0 +1,775 @@
+{
+  return std::make_unique<RedisCacheAioRequest>(this);  
+}
+
+rgw::AioResultList RedisDriver::get_async(const DoutPrefixProvider* dpp, optional_yield y, rgw::Aio* aio, const std::string& key, off_t ofs, uint64_t len, uint64_t cost, uint64_t id) 
+{
+  rgw_raw_obj r_obj;
+  r_obj.oid = key;
+
+  return aio->get(r_obj, rgw::Aio::cache_read_op(dpp, y, this, ofs, len, key), cost, id);
+}
+
+int RedisDriver::AsyncReadOp::init(const DoutPrefixProvider *dpp, CephContext* cct, const std::string& file_path, off_t read_ofs, off_t read_len, void* arg)
+{
+  ldpp_dout(dpp, 20) << "RedisCache: " << __func__ << "(): file_path=" << file_path << dendl;
+  aio_cb.reset(new struct aiocb);
+  
+  memset(aio_cb.get(), 0, sizeof(struct aiocb));
+  aio_cb->aio_fildes = TEMP_FAILURE_RETRY(::open(file_path.c_str(), O_RDONLY|O_CLOEXEC|O_BINARY));
+
+  if (aio_cb->aio_fildes < 0) {
+      int err = errno;
+      ldpp_dout(dpp, 1) << "ERROR: RedisCache: " << __func__ << "(): can't open " << file_path << " : " << " error: " << err << dendl;
+      return -err;
+  }
+
+  if (cct->_conf->rgw_d3n_l1_fadvise != POSIX_FADV_NORMAL) {
+      posix_fadvise(aio_cb->aio_fildes, 0, 0, g_conf()->rgw_d3n_l1_fadvise);
+  }
+
+  bufferptr bp(read_len);
+  aio_cb->aio_buf = bp.c_str();
+  result.append(std::move(bp));
+
+  aio_cb->aio_nbytes = read_len;
+  aio_cb->aio_offset = read_ofs;
+  aio_cb->aio_sigevent.sigev_notify = SIGEV_THREAD;
+  aio_cb->aio_sigevent.sigev_notify_function = libaio_cb_aio_dispatch;
+  aio_cb->aio_sigevent.sigev_notify_attributes = nullptr;
+  aio_cb->aio_sigevent.sigev_value.sival_ptr = arg;
+
+  return 0;
+}
+
+void RedisDriver::AsyncReadOp::libaio_cb_aio_dispatch(sigval sigval)
+{
+  auto p = std::unique_ptr<Completion>{static_cast<Completion*>(sigval.sival_ptr)};
+  auto op = std::move(p->user_data);
+  const int ret = -aio_error(op.aio_cb.get());
+  boost::system::error_code ec;
+  if (ret < 0) {
+      ec.assign(-ret, boost::system::system_category());
+  }
+
+  ceph::async::dispatch(std::move(p), ec, std::move(op.result));
+}
+
+template <typename Executor1, typename CompletionHandler>
+auto RedisDriver::AsyncReadOp::create(const Executor1& ex1, CompletionHandler&& handler)
+{
+  auto p = Completion::create(ex1, std::move(handler));
+  return p;
+}
+
+template <typename ExecutionContext, typename CompletionToken>
+auto RedisDriver::get_async(const DoutPrefixProvider *dpp, ExecutionContext& ctx, const std::string& key,
+                off_t read_ofs, off_t read_len, CompletionToken&& token)
+{
+  namespace net = boost::asio;
+  using boost::redis::connection;
+  using boost::redis::request;
+  using boost::redis::response;
+
+  std::string location = partition_info.location + key;
+
+  ldpp_dout(dpp, 20) << "RedisCache: " << __func__ << "(): location=" << location << dendl;
+
+  request req;
+  response< std::map<std::string, std::string> > resp;
+
+  req.push("HGETALL", key);
```

Comment:
is [HGETALL](https://redis.io/commands/hgetall/) the right operation to read object data? i think that made sense for the directory, where you store a map of attributes for each object entry. but if we're just storing object data, could [GETRANGE](https://redis.io/commands/getrange/) work instead?

Reply:
That's a good point; I assumed that `get_async` was analogous to a regular `get` call which returns everything, but its usage is only for retrieving object data. GETRANGE seems like it would work well within the context of chunking.